Create a RangeSlider widget

ianhi opened the pull request:
## PR Summary
Closes: https://github.com/matplotlib/matplotlib/issues/18563

Creates a SliderWidget that has a movable min and movable max. This enables doing things like this:
![range_slider](https://user-images.githubusercontent.com/10111092/97388179-f1331e00-18ad-11eb-9850-bd5c5a02c887.gif)

The `RangeSlider` is as analogous to `Slider` with the major difference being that `val` is an array rather than a float. To accomplish this I made a new class `_SliderBase` that both the sliders inherit their callback and other shared functionality from. 

## PR Checklist

<!-- Please mark any checkboxes that do not apply to this PR as [N/A]. -->

- [x] Has pytest style unit tests (and `pytest` passes).
- Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
   - [x] code and tests
   - [x] example - keeps complaining about the import for the references
- [x] New features are documented, with examples if plot related.
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [x] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [x] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [N/A] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Do not create the PR out of master, but out of a separate branch.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

ianhi commented on examples/widgets/range_slider_demo.py:
```diff
@@ -0,0 +1,44 @@
+# The use of the following functions, methods, classes and modules is shown
+# in this example:
+
+import matplotlib
```
How am I supposed to make flake8 not complain about this? I don't see any substantive difference from the file I borrowed how to make this reference from.

QuLogic replied:
Add an `E402` entry to `.flake8`.

QuLogic commented on lib/matplotlib/widgets.py:
```diff
@@ -485,21 +514,290 @@ def on_changed(self, func):
+        val[1] = self._max_in_bounds(val[1])
+        xy = self.poly.xy
+        if self.orientation == "vertical":
+            xy[4] = 0, val[0]
```
I don't understand why 4 comes before 0?

ianhi replied:
I found this more intuitive when writing this because it's the same as the zero point, so i wrapped it around to the front of the order. I'll push it back to the end

QuLogic commented on lib/matplotlib/widgets.py:
```diff
@@ -485,21 +514,290 @@ def on_changed(self, func):
+            xy[2] = val[1], 1
+            xy[3] = val[1], 0
+        self.poly.xy = xy
+        fmted = self._format(val)
+        self.valtext.set_text(rf"({fmted[0]}, {fmted[1]})")
```
Again, already a tuple.
```suggestion
        self.valtext.set_text(str(self._format(val)))
```

ianhi commented on lib/matplotlib/widgets.py:
```diff
@@ -485,21 +514,290 @@ def on_changed(self, func):
+        """
+        if min <= self.valmin:
+            if not self.closedmin:
+                return
```
Returing `None` is what the original slider implementation does in this situation. Which is maybe a mistake? Original slider has this code:
```python
        if val <= self.valmin:
            if not self.closedmin:
                return
            val = self.valmin
```
which I think should be changed to 
```python
        if val < self.valmin:
            val = self.valmin
```

which is a behavior change - but also a bug fix. What do you think?

ianhi replied:
oooh I see why this implemented this way. If the user didn't pass in `valstep` but set `closedmin=False` then it is not obvious what value the slider should be set to so it falls back to leaving it at the current value.

ianhi replied:
The correct behavior is returning the current min or current max as appropriate - but definitely not returning None.

Thanks!

ianhi commented on lib/matplotlib/widgets.py:
```diff
@@ -485,21 +514,290 @@ def on_changed(self, func):
+                horizontalalignment="center",
+            )
+
+            fmted = self._format(valinit)
+            s = rf"({fmted[0]}, {fmted[1]})"
+            self.valtext = ax.text(
+                0.5,
+                -0.02,
+                s,
```
I don't totally know why this happens but if you use `str(self._format(valinit))` something totally breaks down and the raw string works around that. This is the error:

```
Unknown symbol: \, found '\'  (at char 0), (line:1, col:1)
Exception in Tkinter callback
Traceback (most recent call last):
  File "/home/ian/Documents/jupyter/matplotlib/lib/matplotlib/_mathtext.py", line 2260, in parse
    result = self._expression.parseString(s)
  File "/home/ian/anaconda3/envs/mpl-dev/lib/python3.8/site-packages/pyparsing.py", line 1955, in parseString
    raise exc
  File "/home/ian/Documents/jupyter/matplotlib/lib/matplotlib/_mathtext.py", line 2418, in unknown_symbol
    raise ParseFatalException(s, loc, "Unknown symbol: %s" % c)
pyparsing.ParseFatalException: Unknown symbol: \, found '\'  (at char 0), (line:1, col:1)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/ian/anaconda3/envs/mpl-dev/lib/python3.8/tkinter/__init__.py", line 1883, in __call__
    return self.func(*args)
  File "/home/ian/anaconda3/envs/mpl-dev/lib/python3.8/tkinter/__init__.py", line 804, in callit
    func(*args)
  File "/home/ian/Documents/jupyter/matplotlib/lib/matplotlib/backends/_backend_tk.py", line 190, in idle_draw
    self.draw()
  File "/home/ian/Documents/jupyter/matplotlib/lib/matplotlib/backends/backend_tkagg.py", line 9, in draw
    super().draw()
  File "/home/ian/Documents/jupyter/matplotlib/lib/matplotlib/backends/backend_agg.py", line 406, in draw
    self.figure.draw(self.renderer)
  File "/home/ian/Documents/jupyter/matplotlib/lib/matplotlib/artist.py", line 74, in draw_wrapper
    result = draw(artist, renderer, *args, **kwargs)
  File "/home/ian/Documents/jupyter/matplotlib/lib/matplotlib/artist.py", line 51, in draw_wrapper
    return draw(artist, renderer, *args, **kwargs)
  File "/home/ian/Documents/jupyter/matplotlib/lib/matplotlib/figure.py", line 2836, in draw
    mimage._draw_list_compositing_images(
  File "/home/ian/Documents/jupyter/matplotlib/lib/matplotlib/image.py", line 133, in _draw_list_compositing_images
    a.draw(renderer)
  File "/home/ian/Documents/jupyter/matplotlib/lib/matplotlib/artist.py", line 51, in draw_wrapper
    return draw(artist, renderer, *args, **kwargs)
  File "/home/ian/Documents/jupyter/matplotlib/lib/matplotlib/_api/deprecation.py", line 432, in wrapper
    return func(*inner_args, **inner_kwargs)
  File "/home/ian/Documents/jupyter/matplotlib/lib/matplotlib/axes/_base.py", line 2838, in draw
    mimage._draw_list_compositing_images(renderer, self, artists)
  File "/home/ian/Documents/jupyter/matplotlib/lib/matplotlib/image.py", line 133, in _draw_list_compositing_images
    a.draw(renderer)
  File "/home/ian/Documents/jupyter/matplotlib/lib/matplotlib/artist.py", line 51, in draw_wrapper
    return draw(artist, renderer, *args, **kwargs)
  File "/home/ian/Documents/jupyter/matplotlib/lib/matplotlib/text.py", line 669, in draw
    bbox, info, descent = textobj._get_layout(renderer)
  File "/home/ian/Documents/jupyter/matplotlib/lib/matplotlib/text.py", line 312, in _get_layout
    w, h, d = renderer.get_text_width_height_descent(
  File "/home/ian/Documents/jupyter/matplotlib/lib/matplotlib/backends/backend_agg.py", line 235, in get_text_width_height_descent
    self.mathtext_parser.parse(s, self.dpi, prop)
  File "/home/ian/Documents/jupyter/matplotlib/lib/matplotlib/mathtext.py", line 452, in parse
    return self._parse_cached(s, dpi, prop, _force_standard_ps_fonts)
  File "/home/ian/Documents/jupyter/matplotlib/lib/matplotlib/mathtext.py", line 473, in _parse_cached
    box = self._parser.parse(s, font_output, fontsize, dpi)
  File "/home/ian/Documents/jupyter/matplotlib/lib/matplotlib/_mathtext.py", line 2262, in parse
    raise ValueError("\n".join(["",
ValueError: 
\\mathdefault{0.25}
^
```

ianhi replied:
I can't figure out a less weird way around the error than the raw string trickery, but I did move the complexity of the raw string into the `_format` function so it's all in one place.

QuLogic replied:
Ah, I made a mistake; casting a tuple to a string directly will use the repr of the contents, making the interior strings into quoted and escaped versions of themselves.

However, there's no need to use a raw string; that only has an effect on the literal value, not what's interpolated via f-strings:
```python
In [1]: a = r'\mathdefault{123}'
In [2]: b = r'\mathdefault{987}'

In [3]: literal = f'({a}, {b})'
In [4]: raw_literal = rf'({a}, {b})'
In [5]: cast = str((a, b))

In [6]: literal
Out[6]: '(\\mathdefault{123}, \\mathdefault{987})'

In [7]: raw_literal
Out[7]: '(\\mathdefault{123}, \\mathdefault{987})'

In [8]: cast
Out[8]: "('\\\\mathdefault{123}', '\\\\mathdefault{987}')"

In [9]: literal == raw_literal
Out[9]: True
```

ianhi replied:
oh interesting. Thanks! 

Fixed now

ianhi commented on lib/matplotlib/widgets.py:
```diff
@@ -485,21 +514,290 @@ def on_changed(self, func):
+        if self.valfmt is not None:
+            return (self.valfmt % val[0], self.valfmt % val[1])
+        else:
+            s1 = super()._format(val[0])
+            s2 = super()._format(val[1])
+            # fmt.get_offset is actually the multiplicative factor, if any.
+            return (s1 + self._fmt.get_offset(), s2 + self._fmt.get_offset())
```
I think it is saved by the `slidermin` and `slidermax`. It would be nice to pass both values simultaneously to `format_ticks` so I'll just make `_format` not part of the superclass. There has to be same branching on `self.valfmt` in RangeSlider anyway so theres not much loss in each slider implementing `_format`.

QuLogic commented on lib/matplotlib/widgets.py:
```diff
@@ -485,21 +513,289 @@ def on_changed(self, func):
+    def _format(self, val):
+        """Pretty-print *val*."""
+        if self.valfmt is not None:
+            return (self.valfmt % val[0], self.valfmt % val[1])
```
This should be made a string too now, to match the `else`?

ianhi replied:
indeed - done

ianhi commented:
Rebased and added more explanatory text to the example + added a bit more functionality. The example now looks like this:

![range-slider](https://user-images.githubusercontent.com/10111092/101114478-9200b180-35af-11eb-81d2-72a5c7ce9ac8.gif)

I also tried out all the Slider examples and they still behave as expected.

QuLogic commented on lib/matplotlib/widgets.py:
```diff
@@ -497,21 +523,290 @@ def on_changed(self, func):
+        if self.valfmt is not None:
+            return f"({self.valfmt % val[0]}, {self.valfmt % val[1]})"
+        else:
+            # fmt.get_offset is actually the multiplicative factor, if any.
```
Duplicated comment.

QuLogic reviewed (commented):
Just some docs are missing.

ianhi commented on lib/matplotlib/widgets.py:
```diff
@@ -229,7 +229,85 @@ def disconnect(self, cid):
+    The base class for constructing Slider widgets. Not intended for direct
+    usage.
+
+    For the slider to remain responsive you must maintain a reference to it.
+    Call :meth:`on_changed` to connect to the slider event.
```
This line is the cause of the doc failutres because `SliderBase` doesn't have an `on_changed` method. Is it preferable to just remove the `meth` directive or to remove the whole sentence?

ianhi replied:
I removed it - The fact that `SliderBase` doesn't have an `on_changed` method probably should mean it's not responsible for documenting that it has that method...

ianhi commented:
Is there any chance of this going into 3.4?

QuLogic commented:
It needs a second reviewer.

QuLogic commented:
I just noticed that vertical is not tested; could you add tests that also check the other orientation? Depending on how things works, it may just be sufficient to parameterize the test for both.

ianhi commented:
> I just noticed that vertical is not tested; could you add tests that also check the other orientation?

Good catch. I added a pytest parametrize to also test the vertical orientation, as well I added that the polygon's points are what we'd expect.

timhoffm commented:
Thanks for the contribution! :+1:
